launcher/internal/oci/cosign/signature.go

```diff
@@ -0,0 +1,64 @@
+// Package cosign contains functionalities to interact with signatures generated by cosign.
+// https://github.com/sigstore/cosign/blob/main/specs/SIGNATURE_SPEC.md.
+package cosign
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	"github.com/google/go-tpm-tools/launcher/internal/oci"
+	"github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// Sig implements oci.Signature interface for cosign-gernated signatures.
+type Sig struct {
+	// Layer represents a layer descriptor for OCI image manifest.
+	Layer v1.Descriptor
+	// Blob represents the opaque data uploaded to OCI registory associated with the layer.
+	Blob []byte
+}
+
+// CosignSigKey is the key of the cosign-generated signature embedded in OCI image manifest.
+const CosignSigKey = "dev.cosignproject.cosign/signature"
+
+var (
+	// Verify that our Sig struct implements the expected public interface.
+	_        oci.Signature = Sig{}
+	encoding               = base64.StdEncoding
+)
+
+// Payload implements oci.Signature interface.
+func (s Sig) Payload() ([]byte, error) {
+	if digest.FromBytes(s.Blob) != s.Layer.Digest {
+		return nil, errors.New("an unmatched payload digest is paired with a layer descriptor digest")
+	}
+	return s.Blob, nil
+}
+
+// Base64Encoded implements oci.Signature interface.
+func (s Sig) Base64Encoded() (string, error) {
+	sig, ok := s.Layer.Annotations[CosignSigKey]
+	if !ok {
+		return "", errors.New("cosign signature not found in the layer annotations")
+	}
+	if _, err := encoding.DecodeString(sig); err != nil {
+		return "", fmt.Errorf("invalid base64 encoded signature: %w", err)
+	}
+	return sig, nil
+}
+
+// PublicKey implements oci.Signature interface.
+// Since public key is attached to the `optional` field of payload, we don't actually implement this method.
```

Comment:
To confirm, given this is under `internal`, this is not meant to be used by users, correct? Just wondering if these should be implemented.

Reply:
Correct. The confidential VM will be the only caller of this method, so we want to make it usable by other packages such as `signaturediscovery` in the library, but NOT export it to users of the library. For the current implementations since the container launcher will send trusted payload with public key attached directly to the service for verification, this method is intended as not implemented.